File: clients/python/baml_core/_impl/deserializer/raw_wrapper/loader.py
import json5 as json  # type: ignore
import re
import typing
from .raw_wrapper import RawWrapper
from .primitive_wrapper import RawBaseWrapper, RawStringWrapper, RawNoneWrapper
from .list_wrapper import ListRawWrapper
from .dict_wrapper import DictRawWrapper
from ..diagnostics import Diagnostics, DeserializerError

def from_string(val: str, diagnostics: Diagnostics) -> RawWrapper:
    return __from_value(val, diagnostics)


def __from_value(val: typing.Any, diagnostics: Diagnostics) -> RawWrapper:
    if val is None:
        return RawNoneWrapper()
    if isinstance(val, bool):
        return RawBaseWrapper(val)
    if isinstance(val, int):
        return RawBaseWrapper(val)
    if isinstance(val, float):
        return RawBaseWrapper(val)
    if isinstance(val, str):
        # First remove any whitespace
        str_val = val.strip()
        # Remove any starting and ending quotes
        if str_val.startswith('"') and str_val.endswith('"'):
            str_val = str_val[1:-1]
        # Remove any starting and ending quotes
        if str_val.startswith("'") and str_val.endswith("'"):
            str_val = str_val[1:-1]

        if str_val.lower() == "true":
            return RawBaseWrapper(True)
        if str_val.lower() == "false":
            return RawBaseWrapper(False)

        is_number = re.match(r"^(\+|-)?\d+(\.\d+)?$", str_val)
        if is_number:
            if "." in str_val:
                return RawBaseWrapper(float(str_val))
            return RawBaseWrapper(int(str_val))

        is_list = str_val.startswith("[") and str_val.endswith("]")
        if is_list:
            try:
                parsed_list = typing.cast(typing.List[typing.Any], json.loads(str_val))
            except:
                parsed_list = None
            if parsed_list:
                return ListRawWrapper([__from_value(item, diagnostics=diagnostics) for item in parsed_list])
        is_dict = str_val.startswith("{") and str_val.endswith("}")
        if is_dict:
            try:
                parsed_obj = typing.cast(
                    typing.Mapping[typing.Any, typing.Any], json.loads(str_val)
                )
            except:
                parsed_obj = None
            if parsed_obj:
                return DictRawWrapper(
                    {__from_value(k, diagnostics=diagnostics): __from_value(v, diagnostics=diagnostics) for k, v in parsed_obj.items()}
                )

        return RawStringWrapper(str_val)
    if isinstance(val, (list, tuple)):
        return ListRawWrapper([__from_value(item, diagnostics=diagnostics) for item in val])
    if isinstance(val, dict):
        return DictRawWrapper(
            {__from_value(key, diagnostics=diagnostics): __from_value(value, diagnostics=diagnostics) for key, value in val.items()}
        )
    
    diagnostics.push_error(DeserializerError("Unrecognized type: {} in value {}".format(type(val), val)))
    diagnostics.to_exception()

    raise Exception("[unreachable] Unsupported type: {}".format(type(val))) 

File: clients/ts/src/baml_lib/deserializer/raw_wrapper/loader.ts
// Import necessary classes
import { Diagnostics } from '../diagnostics';
import { DictRawWrapper } from './dict_wrapper';
import { ListRawWrapper } from './list_wrapper';
import { RawNoneWrapper, RawBaseWrapper, RawStringWrapper } from './primitive_wrapper';
import { RawWrapper } from './raw_wrapper';
import XRegExp from 'xregexp';

export function fromValue(val: any, diagnostics: Diagnostics): RawWrapper {
    if (val === null) {
        return new RawNoneWrapper();
    }
    if (typeof val === 'boolean') {
        return new RawBaseWrapper(val);
    }
    if (typeof val === 'number') {
        if (Number.isInteger(val)) {
            return new RawBaseWrapper(val);
        } else {
            return new RawBaseWrapper(val);
        }
    }
    if (typeof val === 'string') {
        const strVal = val;

        if (strVal.toLowerCase() === "true") {
            return new RawBaseWrapper(true);
        }
        if (strVal.toLowerCase() === "false") {
            return new RawBaseWrapper(false);
        }

        const isNumber = /^(\+|-)?\d+(\.\d+)?$/.test(strVal);
        if (isNumber) {
            if (strVal.includes(".")) {
                return new RawBaseWrapper(parseFloat(strVal));
            }
            return new RawBaseWrapper(parseInt(strVal));
        }

        let parsedList: any[] | undefined = undefined;
        let parsedObj: { [key: string]: any } | undefined = undefined;

        try {
            const parsed = JSON.parse(strVal);
            if (Array.isArray(parsed)) {
                parsedList = parsed;
            } else if (parsed !== null && typeof parsed === 'object') {
                parsedObj = parsed;
            }
        } catch (e) {
            // Handle error
        }

        if (parsedList !== undefined) {
            return new ListRawWrapper(parsedList.map(item => fromValue(item, diagnostics)));
        }

        if (parsedObj !== undefined) {
            const dict: Map<RawWrapper, RawWrapper> = new Map();
            for (const [key, value] of Object.entries(parsedObj)) {
                dict.set(fromValue(key, diagnostics), fromValue(value, diagnostics));
            }
            return new DictRawWrapper(dict);
        }

        // Further string parsing logic for as_inner, as_obj, as_list goes here...
        // Use regular expressions similar to Python's re.findall to extract relevant parts
        let asInner: RawWrapper | undefined = undefined;
        let asObj: RawWrapper | undefined = undefined;
        let asList: RawWrapper | undefined = undefined;

        // Regex to match json within triple backticks
        let startPos = strVal.indexOf('```json');
        if (startPos >= 0 && strVal.indexOf('```', startPos + 4) >= 0) {
            const jsonMatch = XRegExp.matchRecursive(strVal, '```json', '```');
            if (jsonMatch) {
                if (jsonMatch.length >= 1) {
                    asInner = fromValue(jsonMatch[0], diagnostics);
                    // TODO: Handle multiple matches
                }
            }
        } else {

        }

        // Regex to match object-like structures
        if (!(strVal.startsWith('{') && strVal.endsWith('}'))) {
            const objMatch = XRegExp.matchRecursive(strVal, '{', '}', 'g')
            if (objMatch.length > 1) {
                asList = new ListRawWrapper(objMatch.map(match => fromValue(`{${match}}`, diagnostics)));
            } else if (objMatch.length === 1) {
                asObj = fromValue(`{${objMatch[0]}}`, diagnostics);
            }
        }

        // Regex to match list-like structures
        if (asList === undefined) {
            if (!(strVal.startsWith('[') && strVal.endsWith(']'))) {
                const listMatch = XRegExp.matchRecursive(strVal, '\\[', '\\]', 'g')
                if (listMatch.length === 1) {
                    asList = fromValue(`[${listMatch[0]}]`, diagnostics);
                } else {
                    asList = new ListRawWrapper(listMatch.map(match => fromValue(match, diagnostics)));
                }
            }
        }

        return new RawStringWrapper(strVal, asObj, asList, asInner);
    }
    if (Array.isArray(val)) {
        return new ListRawWrapper(val.map(item => fromValue(item, diagnostics)));
    }
    if (typeof val === 'object' && val !== null) {
        const dict: Map<RawWrapper, RawWrapper> = new Map();
        for (const [key, value] of Object.entries(val)) {
            dict.set(fromValue(key, diagnostics), fromValue(value, diagnostics));
        }
        return new DictRawWrapper(dict);
    }

    diagnostics.pushUnknownError(`Unrecognized type: ${typeof val} in value ${val}`);
    diagnostics.toException();

    throw new Error("[unreachable] Unsupported type: " + typeof val);
}
